### ISSUE TITLE
`random` should be redesigned

### ISSUE THREAD

**Issue by Undeterminant** [fish-shell/fish-shell#2642]
This is a continuation of the discussion in #1272.

Right now, `random` in fish mirrors that of bash and isn't very useful at all. I think that it should be redesigned, to cover common use cases of generating random numbers. This thread should be for discussing the exact desired semantics of the function. The goal would be to make a random number function which is cross-platform and easy to use in scripts.

A few requirements:
- <del>It should have flags to designate whether or not the generated number is cryptographically secure. I'm in favour of making it cryptographically secure by default, and people who would rather have speed can choose otherwise.</del>
- <del>There should be an option to generate a uniform random rational number over [0, 1), to a number of decimal places we decide.</del>
- There should be an option to generate a uniform random integer inside either [0, N) or [N, M).
- There should be an option to pick a random item (or number of items) from a list of arguments.
- <del>There should be an option to generate a random UUID.</del>

Things that are covered by other tools (but may want to be included):
- Shuffling or picking random items from files (`shuf` does that well, and technically you could use the above builtin with a `(cat)`).

If we like, we could split these into `random` and `choose` (<del>and `randnorm` and `randuuid`</del>).

**Comment by Undeterminant** [comment]
Sample command interface in docopt format:

```
randnorm [-h | --help] [(--secure | -s) | (--unsecure | -u)]
randrange START [END] [STEP] [-h | --help] [(--secure | -s) | (--unsecure | -u)]
randchoose [ITEMS]... [-h | --help] [(--secure | -s) | (--unsecure | -u)]
randuuid [-h | --help] [(--secure | -s) | (--unsecure | -u)]
```

**Comment by faho** [comment]
I'm going to be devil's advocate here.

Currently, there is _one_ use of the random function in share/* - psub. Though granted, our four uses of `mktemp` could be replaced with a fish function using random. None of these seem likely to improve much with another random implementation, and none benefit from a cryptographically secure one. All these uses just require a name that isn't used yet, and could try a few times until they get one. That seems like the most common use of randomness in shell scripting.

Given that: What would a "proper" random number generator be good for? Why should it be a shell builtin?

To me it seems like this would be a lot of work (anything with "crypto" in the name always seems like a lot of work to do properly) for little gain.

Of course if someone wants to take this on, I'm not stopping them.

**Comment by ridiculousfish** [comment]
Just the other day I wrote a "pick something for me to watch" script that chooses a random file from a directory. It was surprisingly tricky: I ended up having to install `gshuf`. It's a little weird for the shell to be providing `random`, but this may be a case like `seq` where not all OSes ship with a suitable command, and so fish can fill the hole.

Regarding cryptographically secure, it's hopefully very little extra work to use something like arc4random, BlumBlumShub, or `/dev/urandom`. If it's a lot of work then it's not worth doing as faho says.

**Comment by Undeterminant** [comment]
By cryptographically secure I meant `/dev/urandom`, and as far as I'm aware, there's a windows API call that does something similar with little cost. I think that it's more that the wording that people use is "cryptographically secure" when they really mean "true random" versus pseudo-random. It's more about having fewer regular patterns, than using them for generating keys.

I've found myself that at least the "pick a random item from a list" use case is useful to have in a shell. Sure, you can do this with shuf, but that's part of the GNU utils and not necessarily part of any shell.

**Comment by krader1961** [comment]
Thank you, @Undeterminant, for a thoughtfully worded and detailed proposal. Something I don't see often enough. My first reaction involves your recommendation to make the function cryptographically secure but providing an option for "people who would rather have speed can choose otherwise". We're talking about shell scripts. If speed is a concern you should not be using fish. So that level of configurability is a non-starter as far as I'm concerned.

I agree with what @faho wrote. I have no objection to seeing the range of values extended from 15 to 31 bits to reflect the fact very few machines (outside of specialized embedded applications) use a 16 bit integer size today. That would make collisions less likely. I also think there is some merit to adding an option to return a real number rather than an integer. The other proposals to add options to allow, for example, selecting the range of the random integer while trivial to implement can be achieved via other mechanisms. Also, every such option adds complexity and cognitive costs and is therefore something of a slippery slope. Is the feature likely to be useful to a sufficient number of users to justify those costs?

P.S., I know for a fact the original Bourn shell (sh), csh, and tcsh of the 1980's did not support this feature (as someone who was using them at that time). I'm pretty confident that ksh88 (predating the first bash release by a year or two) was the first UNIX shell to provide it and bash copied it. I have a copy of the book "The Kornshell: Command and Programming Language" by David Korn and Morris Bolsky. It documents the feature but provides no rationale or examples for its use.

**Comment by ridiculousfish** [comment]
Just read @Undeterminant's well-considered proposal. Thanks for taking the time to think about it and write it up!

I agree with @krader1961 that we don't need an option to request a CSPRNG. I've changed my mind regarding CSPRNGs: we don't want to even give a hint that fish (or any shell) is the right tool to generate private keys or anything like that. Users should use specialized tools like openssl instead. (Of course we can still use one, just don't say that `random` has crypto-secure output).

Reading from `/dev/urandom` sounds fine to me; the only downside is it may fail if we run out of file descriptors, and so will require annoying error handling. A full user-space implementation may be simpler because it cannot fail. [PCG](http://www.pcg-random.org) is one I've had my eye on for a while.

Regarding the interface:

What does `random norm` do? "Norm" makes me think of a normal distribution, which is different than a uniform distribution.

`random START [END] [STEP]` looks just right to me. I also love the idea of `random choice`. e.g. `open (random choice **.jpg)` to open a random picture; how cool and easy is that!

I think we can omit `random uuid`, since it looks like  Linux, OS X, and BSD all ship with `uuidgen`, which does the job fine.

**Comment by Undeterminant** [comment]
Norm in this case indicates a normal, which is a rather mathematical way of saying between 0 and 1. I've seen the term used before but haven't been too partial with it. The thought process is that the built-in math supports float math, and having the ability to be more precise continuous randomness might be preferred in some scenarios. Although I'm totally fine with dropping it, I just added it because I figured some people might want the format that `Math.random()` provides in Java and JavaScript.

So, modifying the proposal a bit:

* Drop `random norm` and `random uuid`.
* Make sure that the implementation uses a "good" random number generator. Preferably, this will use `/dev/urandom` or a better algorithm than the built-in `rand()` in C. We want to create a truly uniform distribution, which means that stuff like `(random) % N` is out of the question.

How does that sound?

**Comment by faho** [comment]
>How does that sound?

Good!

>The thought process is that the built-in math supports float math

I'm not sure what you mean by that - fish neither has builtin math (it's a function) nor does it support float (because it's a wrapper around bc, which fails at doing modulo and float at the same time). Are you talking about something else?

**Comment by Undeterminant** [comment]
I meant built-in because it's provided by default; you're right that it's not a builtin.

**Comment by YarekTyshchenko** [comment]
Found myself needing one of these things: Pick a random item from a list
```
function random_color
	set list (set_color -c)
	set count (count $list)
	set i (math 'scale=0;'(random)'%'(count $list)'+1')
	echo $list[$i]
end
```

**Comment by occivink** [comment]
I'm going to argue a bit in favor of the proposal. From my own usage, I see the following use cases for a random number generator in the shell:
* Getting a random entry from a variable-size list (see @YarekTyshchenko's post)
* Generating a number in a fixed range, like a virtual dice roller
* Generating a unique identifier while still taking care of collisions, like in `psub`

Of course all of these could be covered by the current `random` implementation (maybe not the third if you need to generate on the order of 10000 identifiers) if you couple it with `math` and modulo but there are some issues with it, which the proposal would solve:
* modulo bias, which skews the distribution of the results
* it's rather annoying to type `(math (random) '%' (count $whatever) +1)` everytime, especially since `%` has special meaning in fish. Looking at my own history, `random` is always used with this kind of construct.

In the end, specifying the range in which you want your number seems like the norm rather than the exception. Even in the third use case I mentioned, you probably have a prior idea of the order of magnitude of how many identifiers are going to be generated and can choose the range accordingly, like two orders above (or 5 orders and completely ignore collisions).

Regarding the complexity and cognitive costs that @krader1961 mentioned, I think the proposed interface is sufficiently simple that there is no ambiguity whatsoever. I think that the `math` construct has more cognitive overhead.

To be honest, I don't really see a use case for the `STEP` value but it's simple enough that I don't have anything against it.

Another argument that I have against the current `random` is regarding the seeding part. I think it's really ugly to be able to modify the internal state of the shell (in a manner that is invisible to the user) and goes against the fish design.

**Comment by krader1961** [comment]
My main issue with your proof of concept change in PR #3602, @occivink, is that it needlessly breaks backward compatibility. It's not clear when we'll have a fish 3.0 release but it is likely to be far in the future. So if you want an improved `random` command it has to be backward compatible.

Too, one of the most compelling use cases for a redesigned `random` command is your first point:

> Getting a random entry from a variable-size list (see @YarekTyshchenko's post)

But your implementation does solve that in an elegant manner. It would be preferable to have `random` itself return a random element using syntax like

```
set -l choice (random choose $values)
```

or alternatively using a flag as in

```
set -l choice (random --pick-one $values)
```

However, it would probably be better to implement a sampling function that allows the user to select "n" unique elements randomly from a sequence; e.g.,

```
set -l choices (random --sample=2 $values)
```

Whether or not to allow the number of items to be omitted, thus defaulting to one, is TBD. And, lastly, before any change is merged we need to specify precisely the ranges that are valid for the start and stop parameters. Do they have to be unsigned? Are the limits going to be 2^32 or 2^64?

> Another argument that I have against the current random is regarding the seeding part.

I don't understand your objection. Anyone using `random` in a script is free to set a new seed using `random $seed`. By definition and necessity the seed is global since we don't have random "objects" like you might have available in a language like Python. Adding a `-s` flag to set the seed does not solve this and only adds more complexity. If you specify a seed every time you call `random` you won't get a random result unless you go out of your way to update the seed on each call. And if you do that I don't see why this idiom isn't acceptable:

```
random $my_seed
set rand_int (random 1 100)
```

**Comment by faho** [comment]
Quick idea: Since there's a compatibility concern, why not make a `choose` builtin, that picks from a list of stuff? If you then wanted a random number, you'd do `choose (seq $start $step $end)`? Then we could even deprecate bare `random` if we wanted.

> I don't understand your objection. Anyone using random in a script is free to set a new seed using random $seed. By definition and necessity the seed is global since we don't have random "objects" like you might have available in a language like Python.

My personal issue with the seeding is that it's waaaay too prominent, and unintuitive as the only argument to `random`. I'd expect it to be a start/end or step value. Being able to set the seed is neat for testing, but pretty much nothing else. This should never be used in production.

**Comment by krader1961** [comment]
> My personal issue with the seeding is that it's waaaay too prominent...

I don't disagree but unless we're going to set a date for a fish 3.0 release we're stuck with the existing semantics if we want to resolve this issue. Also, I don't think the single argument case is especially confusing. If we ignore backward compatibility what does `random 1000` mean? The proof of concept change make that equivalent to `random 0 1000`. Honestly, I don't think requiring specifying the start of the range is onerous enough to warrant breaking backward compatibility.

> why not make a choose builtin...

Because adding a new builtin can cause backward compatibility problems. If fish were more like Python with module and object namespaces this wouldn't be a problem. But that isn't the reality we have to work within. Adding a new builtin is itself something that arguably requires a major release. And in this case isn't needed. It's trivial to augment `random` to behave like a `choose` builtin without ambiguity.

**Comment by occivink** [comment]
I'm okay with augmenting `random` to be closer to the proposal in the short term, but in the long term I still think that it should be sanitized by getting rid of the internal seed.

>Anyone using random in a script is free to set a new seed using random $seed

But by virtue of being global it's easy to subtly get unexpected behavior. What do you do if your script needs to call another script that also sets the seed? Specifying the seed explicitly gets rid of this kind of errors. Also, setting the internal seed once currently causes all your subsequent `random` calls to be deterministic. 
I think seeding is sufficiently uncommon that it's okay to require the user to jump through some hoops.

>select "n" unique elements randomly from a sequence

This is already covered by the likes of `shuf`. I think the problem of picking random elements is a bit higher level than that of simple random number generation which should still be addressed.

>Do they have to be unsigned? Are the limits going to be 2^32 or 2^64?

The proposal suggested `long long` which is what I've implemented.

**Comment by faho** [comment]
>I think seeding is sufficiently uncommon that it's okay to require the user to jump through some hoops.

We're all in agreement on that, the problem is compatibility, which is something we're trying to take more seriously.

Just to be clear, @krader1961, what you propose is that ´random $number` interprets $number as the seed, and `random $number $othernumber` interprets $number as the start and $othernumber as the stop value?

That would be okay. Not great, but considering the added compatibility it's probably the compromise to take.

(There's also an argument to be made that with three arguments the _middle_ one should be the step value for symmetry with `seq`, but I'm not too sure about that since I've never liked it)

>Also, setting the internal seed once currently causes all your subsequent random calls to be deterministic.

That's how these kinds of things tend to work, is it not? I was under the impression that the seed is something you'd use for testing to get deterministic values on purpose.

>>Do they have to be unsigned? Are the limits going to be 2^32 or 2^64?

>The proposal suggested long long which is what I've implemented.

That's apparently also what we use in `test` (try `test 9223372036854775808 -gt 0`). I don't think random numbers larger than that are of any use, and consistency here is nice.

**Comment by occivink** [comment]
>That's how these kinds of things tend to work, is it not? I was under the impression that the seed is something you'd use for testing to get deterministic values on purpose.

Of course, it's the intended usage, but what I mean is that it stays seeded even after you've executed your script and until you exit the shell. 
So it's very much possible for a script to seed my shell "silently" and then I'd be surprised to get the same "random" values (e.g. I have a script to play a random album from my library and it would always play the same sequence).
Having to specify the seed to each `random` invocation is more cumbersome (increment the seed all the time) but prevents this.

Regarding the symmetry with `seq` I think consistency would be preferable actually.

If there's an agreement I can modify my PR to implement the compatible compromise, but I would still like to know what your thoughts are on implementing the full proposal in a future breaking release?

**Comment by krader1961** [comment]
> Just to be clear, @krader1961, what you propose is that...

Exactly. That's required to keep from breaking all existing uses.

@occivink: I still don't see how your introduction of a `-s` flag solves the seeding problem. There is only one PRNG in a given fish instance. As soon as anyone sets a new seed it affects every caller of `random`. If fish had objects this would be trivial to solve. But fish doesn't have objects. Requiring the user to pass a seed on each invocation does not solve the problem and doesn't work because it is incompatible with how PRNGs work; i.e., you'll no longer get a random number sequence.

The only way to solve that problem given the constraints is to implement multiple PRNG states that can be selected by a token. The default being the global state if no token is passed. That is, introduce a `--token`/`-t` flag that takes an arbitrary string. The default token would be "global". Use that to create or lookup a PRNG state using a map. That allows all existing uses of the global PRNG to continue working. Any code that wants a private PRNG it can seed independent of the global one can now do so.

**Comment by occivink** [comment]
I was thinking that if a script needed deterministic rng throughout multiple invocations, they could do something like:
```
set seed 0
function random
	builtin random --seed $seed $argv
	set seed (math $seed +1)
end
```
without relying on any kind of global state.

**Comment by krader1961** [comment]
That won't work, @occivink, as it won't, in general, produce a random sequence. There is a reason it's called a "seed".

**Comment by occivink** [comment]
Ah you're right, it seems like my assumption was incorrect. Your idea of a token looks like a good solution, but at the same time quite complex.

**Comment by faho** [comment]
If there needs to be a "local seed", the obvious solution seems to be a variable - $fish_random_seed?

However, I'm not sure what we're protecting against here - no code you ship or get from someone else should be setting it. However, if you get e.g. a shoddy plugin from some random github repo, that could also set e.g. $PATH to something else or add a function with a common name (make a function called `grep` or `cat` and watch the chaos unfold), which would be _much_ more damaging.

It seems like removing the "only argument is seed" behavior in future would prevent against accidental usage (think of empty variables), and there's no need to protect against purposeful use, because that's stupid.

**Comment by krader1961** [comment]
Also, I forgot to make clear in my previous comment that the "global" PRNG can use a new implementation for generating the sequence but invoking `random` without any arguments (i.e., without specifying a range) must continue to produce an int in the range [0, 32767].

@faho: What @occivink is trying to protect against are other places that call `random` which could affect the sequence seen by a block of code that has explicitly seeded the sequence. In other words consider something like this, obviously contrived, example:

```
random 123  # seed initial state of the PRNG
while true
    set -l randint (random 1 500)
    function_I_do_not_control
    echo $randint
end
```

Where `function_I_do_not_control` might call `random`. If it does then it affects the sequence seen by my `while` block. Worse is if it calls `random` only some of the time it is invoked. Setting the seed as @occivink proposed breaks assumptions baked into the design of a PRNG and therefore doesn't work. And because the proposed change doesn't implement a truly private PRNG object for use by that `while` block setting the seed using either the legacy syntax or the proposed `-s` flag leaks its state to any other piece of code that calls `random`. It's a solution that doesn't work.

In Python this is solved by instantiating a `random.Random()` object wherever you need a private PRNG. Whether or not that is a capability fish needs to support is open for debate. As I outlined earlier the simplest way to do so is to have an internal map that allows us to track multiple PRNG states using a string token. It would be initialized with a single state keyed by the token "global" to support all `random` calls that do not specify a private PRNG token. Obviously this is imperfect because two different pieces of code could use the same token thinking they were the only code in that instance of fish using a given token. But with some token naming guidelines (e.g., use the name of the function that is calling `random` as the token) that risk can be mitigated.

**Comment by faho** [comment]
>Where function_I_do_not_control might call random. If it does then it affects the sequence seen by my while block.

Ah, okay.

>As I outlined earlier the simplest way to do so is to have an internal map that allows us to track multiple PRNG states using a string token.

My idea was that the seed would be stored as a variable, which would then obviously be scoped. The seed itself with the variable lifetime would work as the token (though it's possibly tricky because the RNG would need to listen to the variable state or rewrite it to include the current rng state).

**Comment by krader1961** [comment]
I'm not thrilled with the idea of introducing another magic variable. It will also be moderately complicated to implement. You'll have to modify variable state to include a reference to the PRNG state. Then the `builtin_random()` function has to fetch the innermost (local, function or global) version of the variable. If it already has a reference to PRNG state use that else instantiate a new PRNG object and tie it to the variable state. If there is no such variable in scope then fallback to the global default PRNG state.

An alternative approach that is similar but doesn't require a new magic var is to augment the `block_t` class to include a PRNG state handle. When `random $seed` is run it finds the nearest enclosing block that is a function definition with the same lexical scope as the current block. If one is found its PRNG state handle is initialized if that hasn't already been done. When a random number is generated the same search is done and the PRNG state for that function is used if it exists. Else the global PRNG state is used. But, again, this is going to be complicated to implement. While that itself isn't an argument against doing it this way or with your magic var I think both implementations are going to be hard for most users to grasp.

In short I think that if we really think (semi-)private PRNGs are needed my idea to name them with a string token is the simplest, least intrusive, solution. I also think that seeing `random --name=myprng $seed` followed by `set -l randval (random --name=myprng 1 100)` in a script makes it crystal clear the user is initializing and using a PRNG that is not the global default. Whereas the magic var and scoped `random $seed` behavior is not at all obvious. Note that in this paragraph I've used `--name` rather than `--token`. I think the former is going to be clearer to most users even if we use "token" internally.

**Comment by faho** [comment]
I think we've probably sown enough seeds. I.e. the seed stuff can and should be done later.

So, the design we've come up with is:

- `random SEED` - for backwards compat.
- `random START [STEP] STOP` - where the args are of  type long long (i.e signed integers between -VERYBIGNUMBER and +VERYBIGNUMBER)
- `choose` can be implemented as a function on top (but is more of a bonus)

Any further changes can only be done in a major release.

Okay?

**Comment by krader1961** [comment]
One more point

* `random` (no arguments) returns a value in the range [0, 32767] like it does today.

**Comment by ridiculousfish** [comment]
faho's proposal sounds good to me!

**Comment by krader1961** [comment]
@occivink: Want to rework your PR in light of the consensus we've reached? I only skimmed it but the core of the change to the newly available C++11 functions looked okay. If you rework the rest of it to retain backward compatibility I think we'd be happy to merge it.

**Comment by occivink** [comment]
Sure, I'll do that. To be honest I'm not sure if the code in the pr was actually correct in light of what I'm reading [here](http://www.pcg-random.org/posts/cpp-seeding-surprises.html) but I'll try to fix the issues.

**Comment by krader1961** [comment]
There is no good reason to use the slow, complicated, 64-bit Mersenne Twister PRNG as you did in your original PR, @occivink. The LCG generator `minstd_rand()` is more than adequate for fish. We absolutely do not need or want a RNG suitable for cryptography. See the comments in this issue from December 2015.

**Comment by occivink** [comment]
I think `minstd_rand()` would be inadequate because its period is `2^31-1` and we want to at least be able to produce the full `long long` range. Mersenne Twister may be much more complex, but it's all hidden away anyway.

We can discuss this further in the new PR.

**Comment by ridiculousfish** [comment]
Let's move discussion to the PR #3619

**Comment by krader1961** [comment]
I think the following is important enough to be copied from PR #3619 since many people searching for this won't include GitHub pull requests in their search criteria. The initial comment that is excerpted as a quote is from the author of the pull request justifying why they are using the Mersenne Twister algorithm in their proposed change. That is, the PR author believes we should implement a solution that is suitable for cryptographic and similar applications. The following text is my response in that PR:

> ...somebody assuming that it's enough and getting really inadequate results...

This was discussed extensively. The decision was that we would not implement a `random` suitable for applications needing hard guarantees. People should use tools like openssl where appropriate. See @ridiculousfish's [comment](https://github.com/fish-shell/fish-shell/issues/2642#issuecomment-168079743). As part of this change the man page should be modified to make it crystal clear no one should trust our implementation to be safe for use in cryptographic or equally demanding applications even if we use the Mersenne Twister engine. We do not have the expertise or desire to take on the responsibility for providing such guarantees. This is also why I am opposed to using a 64-bit generator. It makes it too likely someone will think they can safely use it in situations where it is not appropriate.

**Comment by krader1961** [comment]
Fixed by PR #3619 (commits 7996e15ad and 1ace742b6cd).

**Comment by nixpulvis** [comment]
I get that we are following the C++ style here, but at the very least telling the user that a seed was set from `random SEED` would be nice. I was just trying out random (haha) commands and it took me a sec to figure out that the seed was set. The functions `srand` and `rand` make this pretty clear, but here it's a bit less obvious. Small thing, but might be nice.

**Comment by zanchey** [comment]
Come again? Is the problem that `random 123` didn't produce any output?

**Comment by nixpulvis** [comment]
@zanchey Yea re-reading my comment, I could have been much more clear.

It might be nice if `random` output something... Though I could see why you might not like it too. I would personally rather have another function for setting the seed, but that might just be me.

**Comment by krader1961** [comment]
> I would personally rather have another function for setting the seed.

We try to avoid introducing new commands since a) it's another command that has to be implemented and documented, and b) it's another command that can conflict with a function written by a user or external command on their system. Also, we can't really change this now since doing so would break backward compatibility. I'm curious what you expected to happen when you typed `random \<some-int\>`. Also, is there something we can do to make the documentation for `random` clearer?

**Comment by nixpulvis** [comment]
The documentation is really good, and everything makes perfect sense once you read it. I guess I'm just stuck in my ways about wanting the other function for setting the random number. Again personally I'd love the following semantics for `random`.

```fish
random  # [0; 32767]
random END  # [0; END] 
random START END  # [START; END]
...
```

**Comment by krader1961** [comment]
Thank you for the feedback, @nixpulvis. Your suggestion is reasonable and had you provided it when we were discussing this change there is a reasonable chance we would have done it that way. Unfortunately, we can't change it now because doing so would break anyone depending on the existing, documented, interface.

**Comment by nixpulvis** [comment]
@krader1961 yea no problem, I can always define my own. ;)
